Add Jenkins e2e test job for multi-zone Ubernetes Lite.

quinton-hoole opened the pull request:
Work in progress as this still needs to be tested, and won't work until #17919 (Ubernetes Lite support for GCE) is merged.

@justinsb for preliminary review.

quinton-hoole commented:
Fixes #19414 (Add Jenkins test job for multi-zone (Ubernetes Lite) clusters)

justinsb commented on hack/e2e-internal/e2e-up.sh:
```diff
@@ -31,5 +31,14 @@ source "${KUBE_VERSION_ROOT}/cluster/${KUBERNETES_PROVIDER}/util.sh"
+    for KUBE_GCE_ZONE in $E2E_ZONES
+    do
+	KUBE_GCE_ZONE=${KUBE_GCE_ZONE} KUBE_SHARE_MASTER=${KUBE_SHARE_MASTER} "${KUBE_VERSION_ROOT}/cluster/kube-up.sh"
+	KUBE_SHARE_MASTER="true" # For subsequent zones we use the existing master
```
FYI: renamed KUBE_SHARE_MASTER -> KUBE_USE_EXISTING_MASTER, at your suggestion.  That's what you get for making good suggestions :-)

justinsb commented:
Looks good.  Do we release resources correctly?

quinton-hoole commented:
@justinsb Good catch re: the missing kube-down logic.  Doh!  I'll fix that.

Commit:
Add Jenkins e2e test job for multi-zone Ubernetes Lite

quinton-hoole commented:
@justinsb Removed WIP.  PTAL.

quinton-hoole commented:
Oops, that e2e error is legit.  I'll fix the bug in the script:

```
./hack/jenkins/e2e.sh: line 987: E2E_MULTIZONE: unbound variable
```

quinton-hoole commented:
Hopefully fixed by https://github.com/quinton-hoole/kubernetes/commit/7f79ce569ed8e10664b38e2c1b514231b54df442.
Lets see what Jenkins thinks about it.
@k8s-bot test this please

quinton-hoole commented:
@k8s-bot test this please

quinton-hoole commented:
Unit test failure looks like a flake, unrelated to this PR: 
```
All objects updated successfully!!
+++ [0126 21:15:42] Killing api server
+++ [0126 21:15:42] Starting kube-apiserver with KUBE_API_VERSIONS: v1 and runtime-config: api/all=false,api/v1=true
Flag --address has been deprecated, see --insecure-bind-address instead
Flag --public-address-override has been deprecated, see --bind-address instead
Flag --port has been deprecated, see --insecure-port instead
Flag --public-address-override has been deprecated, see --bind-address instead
Flag --kubelet-port has been deprecated, kubelet-port is deprecated and will be removed
I0126 21:15:42.138284    5907 server.go:176] Will report 172.17.0.3 as public IP address.
I0126 21:15:42.138355    5907 plugins.go:71] No cloud provider specified.
I0126 21:15:42.138527    5907 genericapiserver.go:342] Node port range unspecified. Defaulting to 30000-32767.
[restful] 2016/01/26 21:15:42 log.go:30: [restful/swagger] listing is available at https://172.17.0.3:6443/swaggerapi/
[restful] 2016/01/26 21:15:42 log.go:30: [restful/swagger] https://172.17.0.3:6443/swaggerui/ is mapped to folder /swagger-ui/
!!! [0126 21:17:46] Timed out waiting for apiserver:  to answer at http://127.0.0.1:8080/healthz; tried 120 waiting 1 between each
!!! Error in ./hack/test-update-storage-objects.sh:48
  'return 1' exited with status 1
Call stack:
  1: ./hack/test-update-storage-objects.sh:48 startApiServer(...)
  2: ./hack/test-update-storage-objects.sh:123 main(...)
Exiting with status 1
+++ [0126 21:17:46] Killing api server
+++ [0126 21:17:46] Clean up complete
```

@k8s-bot unit test this

quinton-hoole commented:
@justinsb can haz LGTM?  Tested and working now. cc @ihmccreery @ixdy @kubernetes/goog-testing

ixdy commented on cluster/gce/util.sh:
```diff
@@ -1303,7 +1303,17 @@ function test-teardown {
     "${NODE_TAG}-${INSTANCE_PREFIX}-nodeports" || true
-  "${KUBE_ROOT}/cluster/kube-down.sh"
+  if [[ ${MULTIZONE:-} == "true" ]]; then
+      local zones=( $E2E_ZONES )
```
minor bash nit: prefer `${E2E_ZONES}`

quinton-hoole replied:
@ixdy I need an array here rather than a toke list, so that I can iterate through it in reverse order.  Are you OK with that?

ixdy replied:
ah, right. maybe `${E2E_ZONES[*]}`? bash arrays are always mysterious.

(similar nit in e2e-up.sh)

quinton-hoole replied:
as agreed in in-person discussion, that won't work.  Leaving as is.

ixdy replied:
minor nit: gonna go back to my original style rec of s/`$E2E_ZONES`/`${E2E_ZONES}`/, so here
```
local zones=( ${E2E_ZONES} )
```

ixdy commented on cluster/gce/util.sh:
```diff
@@ -1303,7 +1303,17 @@ function test-teardown {
+      done
+      KUBE_GCE_ZONE=${KUBE_GCE_ZONE} "${KUBE_VERSION_ROOT}/cluster/kube-down.sh"
+  else
+      "${KUBE_VERSION_ROOT}/cluster/kube-down.sh"
```
did you intentionally change from `${KUBE_ROOT}` to `${KUBE_VERSION_ROOT}`?

quinton-hoole replied:
@ixdy Ooh - good catch.  Not intentional.  I might have done a dodgy rebase/merge.  Let me check and fix.

ixdy commented on hack/e2e-internal/e2e-up.sh:
```diff
@@ -31,5 +31,14 @@ source "${KUBE_VERSION_ROOT}/cluster/${KUBERNETES_PROVIDER}/util.sh"
+if [[ ${MULTIZONE:-} == "true" ]]; then
+    for KUBE_GCE_ZONE in $E2E_ZONES
+    do
+	KUBE_GCE_ZONE=${KUBE_GCE_ZONE} KUBE_SHARE_MASTER=${KUBE_SHARE_MASTER} "${KUBE_VERSION_ROOT}/cluster/kube-up.sh"
```
where is `KUBE_SHARE_MASTER` set? is this supposed to be `${KUBE_USE_EXISTING_MASTER}`?

quinton-hoole replied:
@ixdy blush.  Yes.

ixdy commented on hack/jenkins/e2e.sh:
```diff
@@ -353,6 +353,26 @@ case ${JOB_NAME} in
 
+  # Runs all non-flaky, non-slow tests on GCE, sequentially,
+  # and in a multi-zone ("Ubernetes Lite") cluster.
+  kubernetes-e2e-gce-ubernetes-lite)
```
will also need to update hack/jenkins/job-configs.

quinton-hoole replied:
@ixdy Thanks will do.

quinton-hoole commented:
@ixdy thanks for the thorough review.

quinton-hoole commented:
@ixdy Comments addressed in commit 91cf292.  PTAL.

ixdy commented on hack/e2e-internal/e2e-up.sh:
```diff
@@ -31,5 +31,14 @@ source "${KUBE_VERSION_ROOT}/cluster/${KUBERNETES_PROVIDER}/util.sh"
+if [[ ${MULTIZONE:-} == "true" ]]; then
+    for KUBE_GCE_ZONE in $E2E_ZONES
+    do
+	KUBE_GCE_ZONE=${KUBE_GCE_ZONE} KUBE_USE_EXISTING_MASTER=${KUBE_USE_EXISTING_MASTER} "${KUBE_VERSION_ROOT}/cluster/kube-up.sh"
```
I think you either need to set `$KUBE_USE_EXISTING_MASTER="false"` or use a default here, as otherwise the first iteration will fail.

ixdy replied:
and as in the other file, quote the variables

ixdy commented on hack/e2e-internal/e2e-up.sh:
```diff
@@ -31,5 +31,14 @@ source "${KUBE_VERSION_ROOT}/cluster/${KUBERNETES_PROVIDER}/util.sh"
 
-"${KUBE_VERSION_ROOT}/cluster/kube-up.sh"
+if [[ ${MULTIZONE:-} == "true" ]]; then
+    for KUBE_GCE_ZONE in $E2E_ZONES
```
same

quinton-hoole replied:
done

ixdy commented on cluster/gce/util.sh:
```diff
@@ -1299,7 +1299,17 @@ function test-teardown {
+      # tear them down in reverse order, finally tearing down the master too.
+      for ((zone_num=${#zones[@]}-1; zone_num>0; zone_num--))
+      do
+	  KUBE_GCE_ZONE=${zones[zone_num]} KUBE_USE_EXISTING_MASTER="true" "${KUBE_ROOT}/cluster/kube-down.sh"
```
maybe put `${zones[zone_num]}` in quotes, per style recs?

quinton-hoole replied:
done

ixdy commented on cluster/gce/util.sh:
```diff
@@ -1299,7 +1299,17 @@ function test-teardown {
+      do
+	  KUBE_GCE_ZONE=${zones[zone_num]} KUBE_USE_EXISTING_MASTER="true" "${KUBE_ROOT}/cluster/kube-down.sh"
+      done
+      KUBE_GCE_ZONE=${KUBE_GCE_ZONE} "${KUBE_ROOT}/cluster/kube-down.sh"
```
same: `${KUBE_GCE_ZONE}`

quinton-hoole replied:
done

ixdy commented:
sorry, this fell off my dashboard since I'm not the assigned reviewer.

LG besides minor style nits, basically. this also needs a rebase (probably trivial)?

quinton-hoole commented:
Thanks @ixdy.  All comments LGTM.  Will do today, rebase and get merged.

quinton-hoole commented:
@ixdy Pushed final nits and rebase.  LGTM?

ixdy commented on cluster/gce/util.sh:
```diff
@@ -1272,7 +1272,17 @@ function test-teardown {
+      do
+	  KUBE_GCE_ZONE="${zones[zone_num]}" KUBE_USE_EXISTING_MASTER="true" "${KUBE_ROOT}/cluster/kube-down.sh"
+      done
+      KUBE_GCE_ZONE="${KUBE_GCE_ZONE}" "${KUBE_ROOT}/cluster/kube-down.sh"
```
it seems that `KUBE_GCE_ZONE` is only set in `hack/jenkins/e2e.sh`, so I'm worried that runs will fail here. (The cluster/gce scripts read `KUBE_GCE_ZONE` but have a default if it's not set.)

quinton-hoole replied:
Good catch!  Done.

ixdy commented:
sorry, should have noticed that the last time around. our env vars are a mess...

quinton-hoole commented:
@ixdy Thanks, I added a default value for KUBE_GCE_ZONE in cluster/gce/util.sh, rebased and squashed.  LGTY?

ixdy commented on cluster/gce/util.sh:
```diff
@@ -1205,7 +1205,17 @@ function test-teardown {
+      do
+	  KUBE_GCE_ZONE="${zones[zone_num]}" KUBE_USE_EXISTING_MASTER="true" "${KUBE_ROOT}/cluster/kube-down.sh"
+      done
+      KUBE_GCE_ZONE="${KUBE_GCE_ZONE:-us-central1-b}" "${KUBE_ROOT}/cluster/kube-down.sh"
```
actually, shouldn't this just be `KUBE_GCE_ZONE="${zones[0]}"`? trying to read `$KUBE_GCE_ZONE` is just going to get you the last value set in the loop, I think, which will be `${zones[1]}`. (maybe also explicitly set `KUBE_USE_EXISTING_MASTER="false"` just to be safe, in case it's set in the environment already?)

(yet again another comment on this line. I'm sorry. too many ongoing reviews for me.)

quinton-hoole replied:
@ixdy You are so right.  Fixing.

quinton-hoole commented:
Thanks @ixdy , addressed your comment https://github.com/kubernetes/kubernetes/pull/19578#discussion_r52396488 and rebased.
I've tried testing this manually but it's very difficult in my dev environment.  I'd suggest we merge this as soon as it passes automated per-PR e2e (in which case it is highly unlikely to cause any trouble), and the remainder of testing can take place in the Jenkins environment (unless you have better suggestions).

ixdy commented on hack/e2e-internal/e2e-up.sh:
```diff
@@ -31,5 +31,16 @@ source "${KUBE_VERSION_ROOT}/cluster/${KUBERNETES_PROVIDER}/util.sh"
 prepare-e2e
 
-KUBE_TEST_DEBUG=y "${KUBE_VERSION_ROOT}/cluster/kube-up.sh"
+"${KUBE_VERSION_ROOT}/cluster/kube-up.sh"
```
bad rebase

quinton-hoole replied:
Oops!  Fixing.

quinton-hoole commented:
@k8s-bot test this issue #18375

ixdy commented:
@quinton-hoole the failure is that you're calling `cluster/kube-up.sh` twice due to a bad rebase, not the issue you linked.

quinton-hoole commented:
@ixdy You are right.  So sorry for the unnecessary interruption.

quinton-hoole commented:
Fix pushed.  Hopefully this one will pass.

quinton-hoole commented:
@ixdy Passed!

ixdy commented:
LGTM, finally! :)
